lib/cusparse/interfaces.jl

```diff
@@ -127,6 +127,73 @@ function Base.:(-)(A::CuSparseMatrix, B::CuSparseMatrixCSR)
     return geam(one(eltype(A)), csrA, -one(eltype(A)), B, 'O')
 end
 
+function Base.reshape(A::CuSparseMatrixCOO{T,M}, dims::NTuple{N,Int}) where {T,N,M}
+    nrows, ncols = size(A)
+    flat_indices = nrows .* (A.colInd .- 1) .+ A.rowInd .- 1
+    new_col, new_row = div.(flat_indices, dims[1]) .+ 1, rem.(flat_indices, dims[1]) .+ 1
+    sparse(new_row, new_col, A.nzVal, dims[1], length(dims) == 1 ? 1 : dims[2], fmt = :coo)
+end
+
+function LinearAlgebra.mul!(Y::CuSparseMatrixCSR{T,M}, A::CuSparseMatrixCSR{T,M}, B::CuSparseMatrixCSR{T,M}) where {T,M}
+    CUSPARSE.version() < v"11.5.1" && throw(ErrorException("This operation is not supported by the current CUDA version."))
+    mm!('N', 'N', one(T), A, B, zero(T), Y, 'O')
+end
+function LinearAlgebra.mul!(Y::CuSparseMatrixCOO{T,M}, A::CuSparseMatrixCOO{T,M}, B::CuSparseMatrixCOO{T,M}) where {T,M}
+    Y2 = CuSparseMatrixCSR(spzeros(T, size(A, 1), size(B, 2)))
+    A2 = CuSparseMatrixCSR(A)
+    B2 = CuSparseMatrixCSR(B)
+    mul!(Y2, A2, B2)
+    copyto!(Y, CuSparseMatrixCOO(Y2))
+end
+function LinearAlgebra.mul!(Y::CuSparseMatrixCSC{T,M}, A::CuSparseMatrixCSC{T,M}, B::CuSparseMatrixCSC{T,M}) where {T,M}
+    Y2 = CuSparseMatrixCSR(spzeros(T, size(A, 1), size(B, 2)))
+    A2 = CuSparseMatrixCSR(A)
+    B2 = CuSparseMatrixCSR(B)
+    mul!(Y2, A2, B2)
+    copyto!(Y, CuSparseMatrixCSC(Y2))
+end
+function LinearAlgebra.:(*)(A::CuSparseMatrixCOO{T,M}, B::CuSparseMatrixCOO{T,M}) where {T,M}
+    Y = CuSparseMatrixCOO(spzeros(T, size(A, 1), size(B, 2)))
+    mul!(Y, A, B)
+end
+
+function SparseArrays.droptol!(A::CuSparseMatrixCOO{T,M}, tol::Real) where {T,M}
+    mask = abs.(A.nzVal) .> tol
+    rows = A.rowInd[mask]
+    cols = A.colInd[mask]
+    datas = A.nzVal[mask]
+    B = sparse(rows, cols, datas, size(A)..., fmt = :coo)
+    copyto!(A, B)
+end
+
+for SparseMatrixType in [:CuSparseMatrixCSC, :CuSparseMatrixCSR]
+    @eval begin
+        Base.reshape(A::$SparseMatrixType{T,M}, dims::NTuple{N,Int}) where {T,N,M} = $SparseMatrixType( reshape(CuSparseMatrixCOO(A), dims) )
+
+        function LinearAlgebra.:(*)(A::$SparseMatrixType{T,M}, B::$SparseMatrixType{T,M}) where {T,M}
+            Y = $SparseMatrixType(spzeros(T, size(A, 1), size(B, 2)))
+            mul!(Y, A, B)
+        end
```

Comment:
```julia
function LinearAlgebra.:(*)(A::CuSparseMatrixCSC{T,M}, B::CuSparseMatrixCSC{T,M}) where {T,M}
    A2 = CuSparseMatrixCSR(A)
    B2 = CuSparseMatrixCSR(B)
    AB = gemm('N', 'N', one(T), A2, B2, 'O')
    CuSparseMatrixCSC(AB)
end

function LinearAlgebra.:(*)(A::CuSparseMatrixCSR{T,M}, B::CuSparseMatrixCSR{T,M}) where {T,M}
     gemm('N', 'N', one(T), A, B, 'O')
end
```

Reply:
If I'm correct, and we should use `mm!`, I think that a non in-place version does not exist.

Reply:
https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cusparse/generic.jl#L233-L355